rclcpp/src/rclcpp/node.cpp

```diff
@@ -605,3 +605,10 @@ Node::get_node_options() const
 {
   return this->node_options_;
 }
+
+void Node::for_each_callback_group(
+  const node_interfaces::NodeBaseInterface::CallbackGroupFunction & func)
+{
+  auto node_base = std::dynamic_pointer_cast<rclcpp::node_interfaces::NodeBase>(node_base_);
```

Comment:
this is pretty unsafe, we're not even checking if the resulting ptr is not null.
We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.
Of course, that only works if all callers are using the new interface, but that's also a problem if we add a method instead of a free function.

Reply:
Will add a check to make sure the pointer is not NULL.

I had originally planned for a free function, but I discussed with @clalancette  and he suggested we make a non-virtual method in ```NodeBase``` and typecast ```NodebaseInterface``` members and then use the ```for_each_callback_group``` to make the code is consistent between rolling and the backport. Effectively the free function would do the same, typecast NodebaseInterface to NodeBase and then implement the for_each routine.

Reply:
> We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.

Yeah, as @adityapande-1995 points out, I don't think it matters whether we make it a free function or a non-virtual member function.  Either case preserves ABI; making it a non-virtual member function is slightly nicer in that you get `this` for free (rather than having to pass it in).

Reply:
yes, the only difference is that the free function alternative doesn't need dynamic casting, as it can receive a reference to the base class directly.

I don't think we're extending `NodeBaseInterface` in another place, so I don't mind much.

Reply:
example:

```cpp
void
for_each_calback_group_thread_safe(NodeBaseInterface & base_interface, FunctionCgT fn) {
  auto & mutex = get_cg_mutex_for_base_interface(&base_interface);
  std::lock_guard guard{mutex};
  for (const auto & cg : base_interface->get_callback_groups()) {
    fn(cg);
  }
}
```

In that case no dynamic cast is needed because [get_callback_groups](https://github.com/ros2/rclcpp/blob/e6733bf2c6cd40c8d8d26ec451042247577f0901/rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp#L131) is already a virtual method.

In that case, we will need the `NodeBaseInterface` constructor to register the mutex though (I think now `NodeBase` is doing that).

Reply:
Ah, I see.  Yeah, that could work more nicely, leveraging the `get_callback_groups` that has to stay here.  @adityapande-1995 this is definitely worth exploring, as getting rid of the `dynamic_cast` will be a lot safer.

Reply:
Will do

Reply:
Added free function, removed the non virtual one and dynamic pointer casts